ts/src/main/scala/ch/epfl/ts/traders/Trader.scala

```diff
@@ -23,15 +41,64 @@ trait TraderCompanion {
   type Key = String
   
   /**
+   * Abstract type which needs to be set by companions to the
+   * concrete Trader class they accompany
+   */
+  type ConcreteTrader <: Trader
+  /**
+   * Needed to overcome type erasure.
+   * @example `override protected val concreteTraderTag = scala.reflect.classTag[MadTrader]`
+   */
+  protected implicit def concreteTraderTag: ClassTag[ConcreteTrader]
+  
+  /**
+   * Check that the given parameters are valid with respect to what is declared by this strategy,
+   * and create a new parameterized Trader if it is the case.
+   * Otherwise, throw a an exception.
+   * 
+   * This is the preferred method to instantiate a Trader, as it will perform parameter checking first. 
+   */
+  final def getInstance(uid: Long, parameters: StrategyParameters, name: String)(implicit builder: ComponentBuilder): ComponentRef = {
+    verifyParameters(parameters)
+    getConcreteInstance(builder, uid, parameters, name)
+  }
+  
+  /**
+   * Provide a new instance of the concrete trading strategy using these parameters.
+   * Can be overriden by concrete TraderCompanion, but the generic implementation should be sufficient.
+   */
+  // TODO: need to use the implicit actor system, or the ComponentBuilder or something
+  protected def getConcreteInstance(builder: ComponentBuilder,
+                                    uid: Long, parameters: StrategyParameters,
+                                    name: String) = {
+    builder.createRef(Props(concreteTraderTag.runtimeClass, uid, parameters), name)
+  }
+  
+  /**
    * Parameters for which the user of the strategy *must* provide a value.
    */
-  def requiredParameters: Map[Key, ParameterTrait[_]]
+  def requiredParameters: Map[Key, ParameterTrait]
   
   /**
    * If the user of the strategy doesn't provide
    * Should not overlap with requiredParameters
    */
-  def optionalParameters: Map[Key, ParameterTrait[_]]
+  def optionalParameters: Map[Key, ParameterTrait] = Map()
   
-  def parameters: Map[Key, ParameterTrait[_]] = requiredParameters ++ optionalParameters
+  /**
+   * All (required and optional) parameters
+   */
+  def parameters: Map[Key, ParameterTrait] = requiredParameters ++ optionalParameters
+  
+  /**
+   * Verify that all required parameters (as specified by this TraderCompanion)
+   * have been provided in `parameters`.
+   * Otherwise, throw an exception.
+   */
+  def verifyParameters(parameters: StrategyParameters) = for {
+    p <- requiredParameters
+    key = p._1
+    theType = p._2
+    if(!parameters.hasWithType(key, theType))
```

Comment:
I think it's a rather idiomatic way of doing it. Also, what's nice this way is that if there's no error, the result is an empty sequence. If I moved the `if` after the `yield`, I would have to `yield` some dummy thing in the `else`.